lib/autoupdate/agent/setup.go

```diff
@@ -273,11 +289,198 @@ func (ns *Namespace) Setup(ctx context.Context, path string) error {
+	ns.log.InfoContext(ctx, "Installing SELinux module.")
+
+	if err := ns.checkSELinux(ctx, true); err != nil {
+		return trace.Wrap(err)
+	}
+
+	// Build the SELinux module in a temp dir, and then install it.
+	fcTempl, err := template.New("selinux file contexts").Parse(selinux.FileContextsTemplate())
+	if err != nil {
+		return trace.Wrap(err, "failed to parse file contexts template")
+	}
+
+	tempDir, err := os.MkdirTemp("", "selinux-*")
+	if err != nil {
+		return trace.Wrap(err, "failed to create temporary directory")
+	}
+	defer os.RemoveAll(tempDir)
+
+	fcFile, err := os.Create(filepath.Join(tempDir, "teleport.fc"))
+	if err != nil {
+		return trace.Wrap(err, "failed to create temporary file")
+	}
+
+	// Generate a file specifying the locations of important dirs so SELinux
+	// will allow Teleport SSH to be able to access them.
+	err = fcTempl.Execute(fcFile, filePaths{
+		InstallDir:     ns.Dir(),
+		DataDir:        ns.dataDir,
+		ConfigPath:     ns.configFile,
+		UpgradeUnitDir: versioncontrol.UnitConfigDir,
+	})
+	if err != nil {
+		return trace.Wrap(err, "failed to expand file contexts template")
+	}
+	if err := fcFile.Sync(); err != nil {
+		return trace.Wrap(err, "failed to synchronize file")
+	}
+
+	err = os.WriteFile(filepath.Join(tempDir, "teleport.te"), selinux.ModuleSource(), 0440)
+	if err != nil {
+		return trace.Wrap(err, "failed to write module source file")
+	}
+
+	cmd := localExec{
+		Dir:      tempDir,
+		Log:      ns.log,
+		ErrLevel: slog.LevelDebug,
+		OutLevel: slog.LevelDebug,
+	}
+
+	// Build and install the module, then ensure files are properly labeled.
+	_, err = cmd.Run(ctx, "make", "-f", selinuxMakefile, "teleport.pp")
+	if err != nil {
+		return trace.Wrap(err, "failed to build module")
+	}
+
+	_, err = cmd.Run(ctx, "semodule", "-i", "teleport.pp")
+	if err != nil {
+		return trace.Wrap(err, "failed to install module")
+	}
+
+	if err := ns.createAndLabelDirs(ctx, cmd); err != nil {
+		return trace.Wrap(err)
+	}
+
+	return nil
+}
+
+func (ns *Namespace) createAndLabelDirs(ctx context.Context, cmd localExec) error {
+	// Create the teleport-upgrade and data dirs ahead of time so we
+	// can ensure they exist when we label them.
+	if err := os.MkdirAll(versioncontrol.UnitConfigDir, defaults.DirectoryPermissions); err != nil {
+		ns.log.WarnContext(ctx, "Failed to create teleport-upgrade directory.", errorKey, err)
+	}
+	if err := os.MkdirAll(ns.dataDir, teleport.PrivateDirMode); err != nil {
+		ns.log.WarnContext(ctx, "Failed to create teleport data directory.", errorKey, err)
+	}
+
+	dirsToLabel := []string{
+		filepath.Clean(ns.installDir),
+		ns.dataDir,
+		versioncontrol.UnitConfigDir,
+	}
+	// Create an empty teleport.yaml config file if it doesn't exist, and
+	// only attempt to label it if it exists.
+	if libutils.FileExists(ns.configFile) {
+		dirsToLabel = append(dirsToLabel, filepath.Dir(ns.configFile))
+	} else {
+		confFile, err := os.OpenFile(ns.configFile, os.O_RDONLY|os.O_CREATE, 0o666)
```

Comment:
Probably want to use the correct permissions here -- 0666 was just an example

Reply:
done